src/ray/gcs/gcs_client/service_based_gcs_client.cc

```diff
@@ -39,28 +39,32 @@ Status ServiceBasedGcsClient::Connect(boost::asio::io_service &io_service) {
   gcs_pub_sub_.reset(new GcsPubSub(redis_gcs_client_->GetRedisClient()));
 
   // Get gcs service address.
-  auto get_server_address = [this]() {
+  get_server_address_func_ = [this]() {
     std::pair<std::string, int> address;
     GetGcsServerAddressFromRedis(redis_gcs_client_->primary_context()->sync_context(),
                                  &address);
     return address;
   };
-  std::pair<std::string, int> address = get_server_address();
-
-  auto re_subscribe = [this](bool is_pubsub_server_restarted) {
-    job_accessor_->AsyncResubscribe(is_pubsub_server_restarted);
-    actor_accessor_->AsyncResubscribe(is_pubsub_server_restarted);
-    node_accessor_->AsyncResubscribe(is_pubsub_server_restarted);
-    task_accessor_->AsyncResubscribe(is_pubsub_server_restarted);
-    object_accessor_->AsyncResubscribe(is_pubsub_server_restarted);
-    worker_accessor_->AsyncResubscribe(is_pubsub_server_restarted);
+  std::pair<std::string, int> address = get_server_address_func_();
+
+  re_subscribe_func_ = [this](bool is_pubsub_server_restarted,
+                              const std::pair<std::string, int> &address) {
+    if (address != current_gcs_server_address_) {
+      current_gcs_server_address_ = address;
+      job_accessor_->AsyncResubscribe(is_pubsub_server_restarted);
+      actor_accessor_->AsyncResubscribe(is_pubsub_server_restarted);
+      node_accessor_->AsyncResubscribe(is_pubsub_server_restarted);
+      task_accessor_->AsyncResubscribe(is_pubsub_server_restarted);
+      object_accessor_->AsyncResubscribe(is_pubsub_server_restarted);
+      worker_accessor_->AsyncResubscribe(is_pubsub_server_restarted);
+    }
   };
 
   // Connect to gcs service.
   client_call_manager_.reset(new rpc::ClientCallManager(io_service));
-  gcs_rpc_client_.reset(new rpc::GcsRpcClient(address.first, address.second,
-                                              *client_call_manager_, get_server_address,
-                                              re_subscribe));
+  gcs_rpc_client_.reset(
+      new rpc::GcsRpcClient(address.first, address.second, *client_call_manager_,
+                            get_server_address_func_, re_subscribe_func_));
```

Comment:
I think we no longer need to call `re_subscribe_func_` in `GcsRpcClient`, right?

Reply:
fixed